### PR TITLE
[RFC] Fix Stella, show number of hits

### PR DESCRIPTION
**Pull request by creshal**
* Stella shoots two bullets that hit separately, this means armour penalty (or bonus) applies twice to each of her attacks. 
* Give a readout of how many hits a doll made; this can be useful to compare different echelon formations (rifles especially, RoF vs. FP formations in general)
* 12s DPS breakdown is imo somewhat more useful than 8s for a typical battle length, YMMV, I just added it and kept the 8s

All of this is just badly hacked together, since I've no idea how half of that simulator works. Further tests and/or refactoring is probably a good idea, I'm open for ideas.

### CODE REVIEW

**Comment by umang-p** [comment]
The Stella changes are fine, though iirc each of her shots also has half the damage which means piercing armor is harder for her than it would be otherwise so these changes would make the sim very slightly overestimate her damage (but its negligible so it should be ok to add)

The 12s damage is another good idea. The only reason the sim uses 8 and 20 at the moment is because an older sim used those numbers. You'll find that a lot of the design decisions for this sim were made arbitrarily like that.

Adding the number of shots is another really good idea, but I do think some changes should be made here. Since the sim uses expected value to account for the RNG with hitting or missing shots, I think accuracy/evasion should not be considered when counting the numbers of shots a doll made. imo it should just be the total number of shots attempted, regardless of whether or not they landed as that would help more in comparing different echelons.

File: brainlets/girlsfrontline/static/girlsfrontline/js/sim.js [hunk 1]
```diff
@@ -3357,6 +3361,7 @@ function simulateBattle() {
               dmg = dmg * Math.min(doll.battle.targets, enemy.count);
             }
           }
+          doll.shots += Math.floor ((doll.links - doll.battle.busylinks) * shotmultiplier * (doll.battle.acc / (doll.battle.acc + enemy.battle.eva)));
```
Comment (umang-p) [inline brainlets/girlsfrontline/static/girlsfrontline/js/sim.js]: Need to account for `hitCount` when `attackBuff !== undefined` is true and don't need to include `(doll.battle.acc / (doll.battle.acc + enemy.battle.eva)`

File: brainlets/girlsfrontline/static/girlsfrontline/js/sim.js [hunk 2]
```diff
@@ -3384,8 +3389,10 @@ function simulateBattle() {
               extradmg *= sureCrit ? (1 + (doll.battle.critdmg / 100)) : 1 + (doll.battle.critdmg * (doll.battle.crit / 100) / 100);
             }
 
+            doll.shots += sureHit? doll.links - doll.battle.busylinks :  Math.floor ((doll.links - doll.battle.busylinks) * (doll.battle.acc / (doll.battle.acc + enemy.battle.eva)));
```
Comment (umang-p) [inline brainlets/girlsfrontline/static/girlsfrontline/js/sim.js:3397]: same as above

File: brainlets/girlsfrontline/static/girlsfrontline/js/sim.js [hunk 3]
```diff
@@ -3384,8 +3389,10 @@ function simulateBattle() {
...
             if ('hitCount' in extraAttack) {
               extradmg *= $.isArray(extraAttack.hitCount) ? extraAttack.hitCount[extraAttack.level - 1] : extraAttack.hitCount;
+              extrashots *= $.isArray(extraAttack.hitCount) ? extraAttack.hitCount[extraAttack.level - 1] : extraAttack.hitCount;
```
Comment (umang-p) [inline brainlets/girlsfrontline/static/girlsfrontline/js/sim.js:3400]: Looks like these shots are not added to the total maybe

File: brainlets/girlsfrontline/static/girlsfrontline/js/sim.js [hunk 4]
```diff
@@ -3679,6 +3695,7 @@ function simulateBattle() {
           }
           dmg *= enemy.battle.vulnerability;
           dmg *= doll.battle.busylinks;
+          doll.shots += sureHit? doll.battle.busylinks : Math.floor (doll.battle.busylinks * (doll.battle.acc / (doll.battle.acc + enemy.battle.eva)));
```
Comment (umang-p) [inline brainlets/girlsfrontline/static/girlsfrontline/js/sim.js:3704]: same as above

File: brainlets/girlsfrontline/static/girlsfrontline/js/sim.js [hunk 5]
```diff
@@ -3743,8 +3763,11 @@ function simulateBattle() {
 
           if ('busylinks' in action) {
             dmg *= doll.battle.busylinks;
+            doll.shots += sureHit? doll.battle.busylinks : doll.battle.busylinks * (doll.battle.acc / (doll.battle.acc + enemy.battle.eva));
           } else {
             dmg *= doll.links;
+            doll.shots += doll.links;
```
Comment (umang-p) [inline brainlets/girlsfrontline/static/girlsfrontline/js/sim.js]: Extra line perhaps

**Comment by creshal** [comment]
> Adding the number of shots is another really good idea, but I do think some changes should be made here. Since the sim uses expected value to account for the RNG with hitting or missing shots, I think accuracy/evasion should not be considered when counting the numbers of shots a doll made.

I wanted this to be able to solve the question "do I even get enough hits in to engage all targets?", and for that accuracy needs to be factored in.

You can just set evasion to 0 if you want to compare raw number of shots; actual hit count with accuracy factored in can be wildly different if you look at accuracy-depressing skills like K31's or underlevelled SAR/AUG. Or accuracy-boosting ones like SRS/T5000.

I'll take a look at the other suggestions after work.

**Comment by creshal** [comment]
> though iirc each of her shots also has half the damage which means piercing armor is harder for her than it would be otherwise so these changes would make the sim very slightly overestimate her damage

This should be taken care of already:

```js
dmg = Math.max(shotmultiplier, 
    doll.battle.fp * multiplier 
     + shotmultiplier * Math.min(2, doll.battle.ap - enemy.battle.armor));
```

Minimum damage in general is set to 2 for Stella (to account for 2 hits doing at least 1 damage each), and the AP bonus/penalty is also multiplied by 2 to account for full penetration happening twice, or armour penalty being applied twice.

**Comment by creshal** [comment]
Alright, hitCount and extraShots should be properly factored in now and duplicate lines were removed. As mentioned above I'd like to retain the accuracy calculations, since it's trivial to just set evasion to 0 if you want to know the total shot count.

**Comment by umang-p** [comment]
While it is trivial, the majority of people who use the sim don't care and just take all numbers they see at face value. For them I think they would just assume that "x dmg in y hits" means y hits are the total number of shots fired.

That said, I thought about this more and asked others and I will add it as-is. However, if there is too much confusion about this I'll change it to show total shots fired and move expected number of hits to a tooltip instead.

**Pull request merged**